### PR TITLE
chore(scanner): Return error from Matcher if no timestamp

### PR DESCRIPTION
**Pull request by dcaravel**
## Description

Changes the Scanner V4 integration to return an error when matcher returns a zero timestamp in `GetMetadata`. This change is intended to reduce ambiguity in clients trying to determine if a timestamp was available vs. not (in order to show an 'error' vs. not, for exmaple)

Also considered changing matcher to return an error directly, however if `GetMetadata` is ever expanded to hold additional values, then error would no longer be appropriate.

## Checklist
- [x] Investigated and inspected CI test results
- [X] Unit test and regression tests added
- [x] ~Evaluated and added CHANGELOG entry if required~
- [x] ~Determined and documented upgrade steps~
- [x] ~Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))~

If any of these don't apply, please comment below.

### Here I tell how I validated my change

Tested via central, prior to change:

```
$ curl -ks -H "Authorization: Bearer $TOKEN" "https://$ACS_HOST/v1/integrationhealth/vulndefinitions?component=SCANNER_V4"
{"lastUpdatedTimestamp":"0001-01-01T00:00:00Z"}
```
Post change, before timestamp avail:
```
$ curl -ksS -H "Authorization: Bearer $TOKEN" "https://$ACS_HOST/v1/integrationhealth/vulndefinitions?component=SCANNER_V4"
{"error":"failed to obtain vulnerability definitions information: no timestamp available","code":13,"message":"failed to obtain vulnerability definitions information: no timestamp available","details":[]}
```
Waited for updaters to finish and then re-ran:
```
$ curl -ksS -H "Authorization: Bearer $TOKEN" "https://$ACS_HOST/v1/integrationhealth/vulndefinitions?component=SCANNER_V4"
{"lastUpdatedTimestamp":"2024-01-18T20:55:43Z"}
```

### Reminder for reviewers

In addition to reviewing code here, reviewers **must** also review testing and request further testing in case the
performed one does not seem sufficient. As a reviewer, you must not approve the change until you understand the
performed testing and you are satisfied with it.

### CODE REVIEW

File: pkg/scanners/scannerv4/scannerv4.go [hunk 1]
```diff
@@ -31,6 +32,10 @@ var (
 
 	scanTimeout     = env.ScanTimeout.DurationSetting()
 	metadataTimeout = 1 * time.Minute
+
+	// zeroProtoTimestampFromTime represents the zero value of a proto
+	// timestamp when initialized from the zero time.
+	zeroProtoTimestampFromTime, _ = protoTypes.TimestampProto(time.Time{})
```
Comment (RTann) [inline pkg/scanners/scannerv4/scannerv4.go:38]: I figured we could just use `protoTypes.Timestamp{}`, but that's not equal to this. TIL

File: pkg/scanners/scannerv4/scannerv4.go [hunk 2]
```diff
@@ -148,6 +153,11 @@ func (s *scannerv4) GetVulnDefinitionsInfo() (*v1.VulnDefinitionsInfo, error) {
 		return nil, fmt.Errorf("pulling metadata from matcher: %w", err)
 	}
 
+	lastTms := metadata.GetLastVulnerabilityUpdate()
+	if lastTms == nil || lastTms.Equal(protoTypes.Timestamp{}) || lastTms.Equal(zeroProtoTimestampFromTime) {
+		return nil, errors.New("no timestamp available")
+	}
+
 	return &v1.VulnDefinitionsInfo{
 		LastUpdatedTimestamp: metadata.GetLastVulnerabilityUpdate(),
```
Comment (RTann) [inline pkg/scanners/scannerv4/scannerv4.go:162]: use the variable here, too

File: pkg/scanners/scannerv4/scannerv4.go [hunk 3]
```diff
@@ -148,6 +153,11 @@ func (s *scannerv4) GetVulnDefinitionsInfo() (*v1.VulnDefinitionsInfo, error) {
 		return nil, fmt.Errorf("pulling metadata from matcher: %w", err)
 	}
 
+	lastTms := metadata.GetLastVulnerabilityUpdate()
+	if lastTms == nil || lastTms.Equal(protoTypes.Timestamp{}) || lastTms.Equal(zeroProtoTimestampFromTime) {
+		return nil, errors.New("no timestamp available")
```
Comment (RTann) [inline pkg/scanners/scannerv4/scannerv4.go:158]: How does this look in the UI?

File: pkg/scanners/scannerv4/scannerv4.go (hunk 1 above)
Comment (dcaravel) [reply pkg/scanners/scannerv4/scannerv4.go:38]: Indeed, learned that during testing myself.

File: pkg/scanners/scannerv4/scannerv4.go (hunk 3 above)
Comment (dcaravel) [reply pkg/scanners/scannerv4/scannerv4.go:158]: Before:
image

After:
image

File: pkg/scanners/scannerv4/scannerv4.go (hunk 2 above)
Comment (dcaravel) [reply pkg/scanners/scannerv4/scannerv4.go]: https://github.com/stackrox/stackrox/pull/9462/commits/5e574092420596352a600bc91f23acf95a75624a

**Review by RTann — APPROVED** [review]

**Pull request merged**